botocore/compress.py

```diff
@@ -0,0 +1,112 @@
+                encoder = getattr(cls, f'_{encoding}_compress_body', None)
+                if encoder is not None:
+                    ce_header = headers.get('Content-Encoding')
+                    if ce_header is None:
+                        headers['Content-Encoding'] = encoding
+                    elif encoding not in ce_header.split(','):
+                        headers['Content-Encoding'] = f'{ce_header},{encoding}'
+                    logger.debug(
+                        'Compressing request with %s encoding', encoding
+                    )
+                    request_dict['body'] = encoder(body)
+                else:
+                    logger.debug(
+                        'Unsupported compression encoding: %s' % encoding
+                    )
+
+    @classmethod
+    def _gzip_compress_body(cls, body):
+        if isinstance(body, str):
+            return gzip.compress(body.encode('utf-8'))
+        elif isinstance(body, (bytes, bytearray)):
+            return gzip.compress(body)
+        elif hasattr(body, 'read'):
+            if hasattr(body, 'seek') and hasattr(body, 'tell'):
+                current_position = body.tell()
+                compressed_obj = cls._gzip_compress_fileobj(body)
+                body.seek(current_position)
+                return compressed_obj
+            return cls._gzip_compress_fileobj(body)
+
+    @staticmethod
+    def _gzip_compress_fileobj(body):
+        compressed_obj = io.BytesIO()
+        with gzip.GzipFile(fileobj=compressed_obj, mode='wb') as gz:
+            while True:
+                chunk = body.read(8192)
+                if not chunk:
+                    break
+                if isinstance(chunk, str):
+                    chunk = chunk.encode('utf-8')
+                gz.write(chunk)
+        compressed_obj.seek(0)
+        return compressed_obj
+
+    @classmethod
+    def _should_compress_request(cls, config, body, operation_model):
+        if (
+            config.disable_request_compression is not True
+            and config.signature_version != 'v2'
+            and operation_model.request_compression is not None
+        ):
+            # Request is compressed no matter the content length if it has a streaming input.
+            # However, if the stream has the `requiresLength` trait it is NOT compressed.
+            if operation_model.has_streaming_input:
+                return (
+                    'requiresLength'
+                    not in operation_model.get_streaming_input().metadata
+                )
+            return (
+                config.request_min_compression_size_bytes
+                <= cls._get_body_size(body)
+            )
+        return False
+
+    @staticmethod
+    def _get_body_size(body):
+        size = determine_content_length(body)
+        if size is None:
+            logger.debug(
+                'Unable to get length of the request body: %s. Not compressing.'
+                % body
```

Comment:
You don't need to use `%` with logger statements. Alternatively, we could use f-strings.
```suggestion
                'Unable to get length of the request body: %s. Skipping compression.', body
```

Reply:
I thought it was encouraged to use %s strings in logging statements for optimization reasons? Its pretty much the only place I ever use them. https://docs.python.org/3/howto/logging.html#optimization

Reply:
Note the difference between:

```python
'Unable to get length of the request body: %s. Not compressing.' % body
```

and

```python
'Unable to get length of the request body: %s. Skipping compression.', body
```

With the former syntax, you're circumventing the logger optimization. We don't want to do that.

Reply:
Ah got it thank you for clarifying!